packages/vitest/src/node/pools/pool.ts

```diff
@@ -0,0 +1,263 @@
+import type { Logger } from '../logger'
+import type { StateManager } from '../state'
+import type { PoolRuntime, PoolTask, WorkerResponse } from './types'
+import { ForksRuntime } from './runtimes/forks'
+import { ThreadsRuntime } from './runtimes/threads'
+import { TypecheckRuntime } from './runtimes/typecheck'
+import { VmForksRuntime } from './runtimes/vmForks'
+import { VmThreadsRuntime } from './runtimes/vmThreads'
+
+const WORKER_START_TIMEOUT = 5_000
+
+interface Options {
+  distPath: string
+  teardownTimeout: number
+  state: StateManager
+}
+
+interface QueuedTask {
+  task: PoolTask
+  resolver: ReturnType<typeof withResolvers>
+  method: 'run' | 'collect'
+}
+
+interface ActiveTask extends QueuedTask {
+  cancelTask: () => Promise<void>
+}
+
+export class Pool {
+  private maxWorkers: number = 0
+  private workerIds = new Map<number, boolean>()
+
+  private queue: QueuedTask[] = []
+  private activeTasks: ActiveTask[] = []
+  private sharedRuntimes: PoolRuntime[] = []
+  private exitPromises: Promise<void>[] = []
+
+  constructor(private options: Options, private logger: Logger) {}
+
+  setMaxWorkers(maxWorkers: number): void {
+    this.maxWorkers = maxWorkers
+
+    this.workerIds = new Map(
+      Array.from({ length: maxWorkers }).fill(0).map((_, i) => [i + 1, true]),
+    )
+  }
+
+  async run(task: PoolTask, method: 'run' | 'collect'): Promise<void> {
+    // Every runtime related failure should make this promise reject so that it's picked by pool.
+    // This resolver is used to make the error handling in recursive queue easier.
+    const testFinish = withResolvers()
+
+    this.queue.push({ task, resolver: testFinish, method })
+    void this.schedule()
+
+    await testFinish.promise
+  }
+
+  private async schedule(): Promise<void> {
+    if (this.queue.length === 0 || this.activeTasks.length >= this.maxWorkers) {
+      return
+    }
+
+    const { task, resolver, method } = this.queue.shift()!
+
+    try {
+      let isMemoryLimitReached = false
+      const runtime = this.getRuntime(task, method)
+
+      const activeTask = { task, resolver, method, cancelTask }
+      this.activeTasks.push(activeTask)
+
+      runtime.on('error', error => resolver.reject(new Error(`[vitest-pool]: Runtime ${task.runtime} emitted error`, { cause: error })))
+
+      async function cancelTask() {
+        await runtime.stop()
+        resolver.reject(new Error('Cancelled'))
+      }
+
+      const onFinished = (message: WorkerResponse) => {
+        if (message?.__vitest_worker_response__ && message.type === 'testfileFinished') {
+          if (task.memoryLimit && message.usedMemory) {
+            isMemoryLimitReached = message.usedMemory >= task.memoryLimit
```

Comment:
Tested memory limit manually, seems to work like before:

<img width="600" src="https://github.com/user-attachments/assets/10a20ac8-5427-43c7-8c01-52beed19f33c" />